diff --git a/hw/arm/bcm2835_peripherals.c b/hw/arm/bcm2835_peripherals.c
index ac153a9..356eca9 100644
--- a/hw/arm/bcm2835_peripherals.c
+++ b/hw/arm/bcm2835_peripherals.c
@@ -98,6 +98,9 @@ static void raspi_peripherals_base_init(Object *obj)
 
     /* UART0 */
     object_initialize_child(obj, "uart0", &s->uart0, TYPE_PL011);
+    
+    /* UART3 */
+    object_initialize_child(obj, "uart3", &s->uart3, TYPE_PL011);
 
     /* AUX / UART1 */
     object_initialize_child(obj, "aux", &s->aux, TYPE_BCM2835_AUX);
@@ -289,6 +292,8 @@ void bcm_soc_peripherals_common_realize(DeviceState *dev, Error **errp)
                 sysbus_mmio_get_region(SYS_BUS_DEVICE(&s->cprman), 0));
     qdev_connect_clock_in(DEVICE(&s->uart0), "clk",
                           qdev_get_clock_out(DEVICE(&s->cprman), "uart-out"));
+    qdev_connect_clock_in(DEVICE(&s->uart3), "clk",
+                          qdev_get_clock_out(DEVICE(&s->cprman), "uart-out"));
 
     memory_region_add_subregion(&s->peri_mr, ARMCTRL_IC_OFFSET,
                 sysbus_mmio_get_region(SYS_BUS_DEVICE(&s->ic), 0));
@@ -338,6 +343,18 @@ void bcm_soc_peripherals_common_realize(DeviceState *dev, Error **errp)
         qdev_get_gpio_in_named(DEVICE(&s->ic), BCM2835_IC_GPU_IRQ,
                                INTERRUPT_AUX));
 
+    /* UART3 */
+    qdev_prop_set_chr(DEVICE(&s->uart3), "chardev", serial_hd(3));
+    if (!sysbus_realize(SYS_BUS_DEVICE(&s->uart3), errp)) {
+        return;
+    }
+    /* yes we reuse INTERRUPT_UART0 */
+    memory_region_add_subregion(&s->peri_mr, UART3_OFFSET,
+                sysbus_mmio_get_region(SYS_BUS_DEVICE(&s->uart3), 0));
+    sysbus_connect_irq(SYS_BUS_DEVICE(&s->uart3), 0,
+        qdev_get_gpio_in_named(DEVICE(&s->ic), BCM2835_IC_GPU_IRQ,
+                               INTERRUPT_UART0));
+
     /* Mailboxes */
     if (!sysbus_realize(SYS_BUS_DEVICE(&s->mboxes), errp)) {
         return;
diff --git a/include/hw/arm/bcm2835_peripherals.h b/include/hw/arm/bcm2835_peripherals.h
index 1eeaeec..fceae28 100644
--- a/include/hw/arm/bcm2835_peripherals.h
+++ b/include/hw/arm/bcm2835_peripherals.h
@@ -59,6 +59,7 @@ struct BCMSocPeripheralBaseState {
     BCM2835PowerMgtState powermgt;
     BCM2835CprmanState cprman;
     PL011State uart0;
+    PL011State uart3;
     BCM2835AuxState aux;
     BCM2835FBState fb;
     BCM2835DMAState dma;
diff --git a/include/hw/arm/raspi_platform.h b/include/hw/arm/raspi_platform.h
index 7bc4807..a2e44f4 100644
--- a/include/hw/arm/raspi_platform.h
+++ b/include/hw/arm/raspi_platform.h
@@ -86,6 +86,7 @@ uint64_t board_ram_size(uint32_t board_rev);
 #define RNG_OFFSET              0x104000
 #define GPIO_OFFSET             0x200000
 #define UART0_OFFSET            0x201000 /* PL011 */
+#define UART3_OFFSET            0x201600 /* PL011 */
 #define MMCI0_OFFSET            0x202000 /* Legacy MMC */
 #define I2S_OFFSET              0x203000 /* PCM */
 #define SPI0_OFFSET             0x204000 /* SPI master */
